### PR TITLE
Split up the backend, frontend, and the docprovider

### LINKED ISSUE #269: Separate out the frontend and the backend
### Problem

[jupyter-collaboration](https://github.com/jupyterlab/jupyter-collaboration) is currently under the [jupyterlab](https://github.com/jupyterlab) organization. It includes the JupyterLab frontend and the [jupyter_server](https://github.com/jupyter-server/jupyter_server) backend. The backend part used to be [jupyter_server_ydoc](https://github.com/jupyter-server/jupyter_server_ydoc), under the [jupyter-server](https://github.com/jupyter-server) organization. I think it was a good choice to separate out the frontend and the backend, since some extensions may want to have access to the shared models without bringing the frontend stuff. Also it would allow [jupyverse](https://github.com/jupyter-server/jupyverse) to not depend (or depend less) on `jupyter-server`.

### Proposed Solution

We could split this repository in two, and have only the JupyterLab frontend stuff in a new repository `jupyterlab/jupyterlab-collaboration`, which would have the backend as a dependency.
The backend part would be in `jupyter-server/jupyter-collaboration`.
What do you think?

cc @krassowski

### PR DESCRIPTION
**Pull request by krassowski**
Closes #269

New structure:
- packages: (npm)
  - `@jupyter/collaboration` (no changes)
  - `@jupyter/collaboration-extension`: filebrowser and server-side execution plugin extracted
  - `@jupyter/docprovider` (no changes)
  - `@jupyter/docprovider-extension` (new) home for filebrowser/drive and server-side execution plugin (for now)
- projects: (python, new)
  - `jupyter-server-ydoc`: jupyter-server extension
  - `jupyter-docprovider`: JupyterLab/Notebook 7+ extension (`@jupyter/docprovider-extension`)
  - `jupyter-collaboration-ui`: JupyterLab/Notebook 7+ extension (`@jupyter/collaboration-extension`)
  - `jupyter-collaboration`: empty package which installs all the three above

### Breaking changes

We should call this a new major version as now it is necessary to import from `jupyter_server_ydoc` rather than `jupyter_collaboration` (though I am not sure if anyone was importing anything as there was no public API before?).

### CODE REVIEW

**Comment by krassowski** [comment]
This is now ready, the release check will be fixed once we make a release. The only question from me is whether it is ok to have linked versioning, or if we want to reset version in the packages which are getting split up and have independent versioning in individual python packages?

File: projects/jupyter-collaboration-ui/pyproject.toml
Comment (davidbrochart) [inline projects/jupyter-collaboration-ui/pyproject.toml:23]:
```suggestion
    "Programming Language :: Python :: 3.11",
    "Programming Language :: Python :: 3.12",
```

File: projects/jupyter-collaboration/pyproject.toml
Comment (davidbrochart) [inline projects/jupyter-collaboration/pyproject.toml:23]:
```suggestion
    "Programming Language :: Python :: 3.11",
    "Programming Language :: Python :: 3.12",
```

File: projects/jupyter-collaboration/pyproject.toml
Comment (davidbrochart) [inline projects/jupyter-collaboration/pyproject.toml:33]: Should we start pinning dependencies? At least we shouldn't install `jupyter-server-ydoc` v0.8.0 which stills exist on [PyPI](https://pypi.org/project/jupyter-server-ydoc).

File: projects/jupyter-docprovider/pyproject.toml
Comment (davidbrochart) [inline projects/jupyter-docprovider/pyproject.toml:23]:
```suggestion
    "Programming Language :: Python :: 3.11",
    "Programming Language :: Python :: 3.12",
```

File: projects/jupyter-server-ydoc/pyproject.toml
Comment (davidbrochart) [inline projects/jupyter-server-ydoc/pyproject.toml:23]:
```suggestion
    "Programming Language :: Python :: 3.11",
    "Programming Language :: Python :: 3.12",
```

**Review by davidbrochart — APPROVED** [review]
Thanks so much Mike, that's quite a lot of changes!

> The only question from me is whether it is ok to have linked versioning, or if we want to reset version in the packages which are getting split up and have independent versioning in individual python packages?

I'd rather not have linked versioning, because IIUC a change released in one package will force a release in the other packages even though nothing has changed in them.
To me that kind of breaks SemVer, right? Imagine a breaking change release in a package where actually nothing has changed, that can be considered a lie :smile: 
I have linked versioning in Jupyverse's plugings, and that also leads to more work with the conda packages.
But what are the drawbacks of independent versioning?

File: projects/jupyter-collaboration/pyproject.toml
Comment (krassowski) [reply projects/jupyter-collaboration/pyproject.toml]: Good point, thanks. I set their versions to 1.0 alpha 0 and pinned here to `>=1.0.0a0`.

**Comment by krassowski** [comment]
Can you move `jupyter-server-ydoc` on PyPI to Project Jupyter org (https://pypi.org/org/jupyter/, same way as jupyter-collaboration already is) and add me as a maintainer?

File: scripts/bump_version.py
Comment (davidbrochart) [inline scripts/bump_version.py:87]: Maybe not use `assert` but a check that raises an exception?

File: RELEASE.md
Comment (davidbrochart) [inline RELEASE.md:10]:
```suggestion
Input appropriate specifier into the `jupyter-releaser` workflow dispatch dialog to bump version numbers for this release.
```

**Review by davidbrochart — COMMENTED** [review]
[Failure](https://github.com/jupyterlab/jupyter-collaboration/actions/runs/8686075463/job/23816832153?pr=280) is unrelated, see https://github.com/jupyter-server/jupyter_server_fileid/issues/75.

**Comment by davidbrochart** [comment]
> Can you move `jupyter-server-ydoc` on PyPI to Project Jupyter org (https://pypi.org/org/jupyter/, same way as jupyter-collaboration already is) and add me as a maintainer?

It seems that I cannot transfer it because I am not an owner or manager of any organizations, so I'll let you do it.
I sent an invitation to `krassowski`, let me know if that was not you PyPI user name.

**Comment by krassowski** [comment]
All green, merging!

**Comment by jtpio** [comment]
> @jtpio can you help with the transfer of `jupyter-server-ydoc`, `jupyter-docprovider` and `jupyter-collaboration-ui` to Jupyter org on PyPI?

Looks like I don't have the rights to do this unfortunately, the `jupyter` org does not show up in the list:

image

Maybe @blink1073 or @fcollonval can help with this.

**Comment by jtpio** [comment]
> We should call this a new major version as now it is necessary to import from `jupyter_server_ydoc` rather than `jupyter_collaboration` (though I am not sure if anyone was importing anything as there was no public API before?).

Could there be a way to shim the imports otherwise? And add a warning to recommend updating the imports. Wondering if that could help avoid making the major version.

**Comment by krassowski** [comment]
> Could there be a way to shim the imports otherwise?

I did look into that. There is only one thing that was exported from the top-level `__init__` which was the `YDocExtension`. However, it was not really useful unless you also imported from other modules and mirroring the hierarchy to ensure imports work did not seem worth it to me.

I am operating on an assumption that no one used the Python API if you know of production uses though then I am happy to revisit.

**Comment by krassowski** [comment]
In the public land I do not see any production imports that would be broken:
- [`from jupyter_collaboration`](
https://github.com/search?q=%22from+jupyter_collaboration%22+NOT+repo%3Ajupyterlab%2Fjupyter-collaboration+NOT+repo%3Akrassowski%2Fjupyter_collaboration+NOT+repo%3AmartinRenou%2Fjupyter-collaboration+NOT+repo%3Ahbcarlos%2Fjupyter-collaboration+NOT+repo%3Afcollonval%2Fjupyter_server_ydoc+NOT+repo%3Adatalayer-externals%2Fjupyter-collaboration+NOT+repo%3Adatalayer-externals%2Fjupyterlab-collaboration&type=code)
- [`import jupyter_collaboration`](https://github.com/search?q=%22import+jupyter_collaboration%22&type=code)

Except for https://github.com/datalayer/jupyter-rtc-test which would need migration CC @echarles.

**Comment by krassowski** [comment]
@blink1073 I invited you as owner to `jupyter-docprovider` and `jupyter-collaboration-ui` on PyPI. Can someone with rights also send an invite for `jupyter-server-ydoc`?

**Comment by davidbrochart** [comment]
I just did.

**Comment by krassowski** [comment]
Thank you all!

**Comment by Zsailer** [comment]
Hey folks, I'm late to the party here. Thanks for doing the work here! I think this is the "right" move going forward.

However, I think we need to keep the 2.x branch open until _at least_ JupyterLab 4.2 goes "stable" (right now it's in RC). This PR raised the minimum requirement for JLab packages to 4.2 beta, forcing folks to use this with pre release of JupyterLab. 

@jzhang20133 and I have been running RTC in production and fixing multiple issues we've observed. We can't depend on an pre-release at this point in time, so we'll need to backport our changes to the 2.x line temporarily and cut releases from there to unblock and continue getting feedback from our users.

> In the public land I do not see any production imports that would be broken:

We have (not yet open-sourced) extensions that import from `jupyter_collaboration`, so this PR also broke our work here. I don't think a shim is necessary—we can tolerate the changes on our end. I think more folks will begin adopting this library though, so we should do this more prudently going forward.

I'll make a 2.x branch, backport our changes, and cut some releases from there.

**Pull request merged**
